cluster_management/src/main/java/com/pinterest/rocksplicator/eventstore/ZkMergeableEventStore.java

```diff
@@ -0,0 +1,238 @@
+/// Copyright 2021 Pinterest Inc.
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+/// http://www.apache.org/licenses/LICENSE-2.0
+
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+
+//
+// @author Gopal Rajpurohit (grajpurohit@pinterest.com)
+//
+
+package com.pinterest.rocksplicator.eventstore;
+
+import com.pinterest.rocksplicator.codecs.Codec;
+import com.pinterest.rocksplicator.codecs.CodecException;
+
+import com.google.common.base.Preconditions;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.locks.InterProcessMutex;
+import org.apache.curator.framework.recipes.locks.Locker;
+import org.apache.curator.framework.state.ConnectionState;
+import org.apache.curator.framework.state.ConnectionStateListener;
+import org.apache.zookeeper.data.Stat;
+
+import java.io.IOException;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+public class ZkMergeableEventStore<R, E> implements MergeableReadWriteStore<R, E> {
+
+  private final AtomicBoolean pathExists;
+  private final String eventHistoryPath;
+  private final String partitionLockPath;
+  private final BatchMergeOperator<R> mergeOperator;
+  private final Supplier<R> zeroSupplier;
+  private final Codec<R, byte[]> recordCodec;
+  private final CuratorFramework zkClient;
+  private final Semaphore ipMutexGuard = new Semaphore(1);
+  private InterProcessMutex interProcessPartitionMutex;
+
+
+  public ZkMergeableEventStore(
+      CuratorFramework zkClient,
+      final String clusterName,
+      final String resourceName,
+      final String partitionName,
+      final Codec<R, byte[]> recordCodec,
+      final Supplier<R> zeroSupplier,
```

Comment:
what is reason to have a zeroSupplier other than just 0?

Reply:
for integer zero is 0. For LeaderEventsHistory it is new LeaderEventsHistory() which has empty list of events...

Reply:
This. store requires zeroSupplier, since it doesn't know what zero means for the template argument.

In our case, you can consider set of all LeaderEventsHistory as a monoid with the merge operator being the associative binary operatorion and identify element being new LeaderEventsHistory().